components/function-controller/internal/controllers/serverless/function_controller.go

```diff
@@ -0,0 +1,760 @@
+								},
+							},
+						},
+						{
+							Name: "runtime",
+							VolumeSource: corev1.VolumeSource{
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{Name: r.config.Build.RuntimeConfigMapName},
+								},
+							},
+						},
+						{
+							Name: "credentials",
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{SecretName: r.config.ImagePullSecretName},
+							},
+						},
+						{
+							Name:         "tekton-home",
+							VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}},
+						},
+						{
+							Name:         "tekton-workspace",
+							VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}},
+						},
+					},
+					InitContainers: []corev1.Container{
+						{
+							Name:    "credential-initializer",
+							Image:   r.config.Build.CredsInitImage,
+							Command: []string{"/ko-app/creds-init"},
+							Args:    []string{fmt.Sprintf("-basic-docker=credentials=http://%s", imageName)},
+							Env: []corev1.EnvVar{
+								{Name: "HOME", Value: "/tekton/home"},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{Name: "tekton-home", ReadOnly: false, MountPath: "/tekton/home"},
+								{Name: "credentials", ReadOnly: false, MountPath: "/tekton/creds-secrets/credentials"},
+							},
+							ImagePullPolicy: corev1.PullIfNotPresent,
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "executor",
+							Image: r.config.Build.ExecutorImage,
+							Args:  []string{fmt.Sprintf("--destination=%s", imageName), "--insecure", "--skip-tls-verify"},
+							Resources: corev1.ResourceRequirements{
+								Limits: corev1.ResourceList{
+									corev1.ResourceMemory: r.config.Build.LimitsMemoryValue,
+									corev1.ResourceCPU:    r.config.Build.LimitsCPUValue,
+								},
+								Requests: corev1.ResourceList{
+									corev1.ResourceMemory: r.config.Build.RequestsMemoryValue,
+									corev1.ResourceCPU:    r.config.Build.RequestsCPUValue,
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{Name: "sources", ReadOnly: true, MountPath: "/src"},
+								{Name: "runtime", ReadOnly: true, MountPath: "/workspace"},
+								{Name: "tekton-home", ReadOnly: false, MountPath: "/tekton/home"},
+							},
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Env: []corev1.EnvVar{
+								{Name: "DOCKER_CONFIG", Value: "/tekton/home/.docker/"},
+							},
+						},
+					},
+					RestartPolicy:      corev1.RestartPolicyNever,
+					ServiceAccountName: r.config.ImagePullAccountName,
+				},
+			},
+		},
+	}
+
+	return job
+}
+
+func (r *FunctionReconciler) buildService(log logr.Logger, instance *serverlessv1alpha1.Function, oldService *servingv1.Service) servingv1.Service {
+	imageName := r.buildExternalImageAddress(instance)
+	annotations := map[string]string{
+		"autoscaling.knative.dev/minScale": "1",
```

Comment:
Please extract whole label names to consts.

Reply:
If the jobs will pass I will merge it as it is and I will change it in PR with integration tests.